core/container/externalbuilder/externalbuilder.go

```diff
@@ -380,23 +384,25 @@ func (b *Builder) Run(ccid, bldDir string, peerConnection *ccintf.PeerConnection
 
 	run := filepath.Join(b.Location, "bin", "run")
 	cmd := b.NewCommand(run, bldDir, launchDir)
-	sess, err := Start(b.Logger, cmd)
+	sess, err := Start(b.Logger, cmd, b.signalChan, b.doneChan)
 	if err != nil {
 		os.RemoveAll(launchDir)
 		return nil, errors.Wrapf(err, "builder '%s' run failed to start", b.Name)
 	}
 
 	go func() {
-		defer os.RemoveAll(launchDir)
+		sess.WaitGroup.Add(1)
 		sess.Wait()
+		os.RemoveAll(launchDir)
+		sess.WaitGroup.Done()
```

Comment:
I don't see how this can work. We're waiting before calling Done in the same synchronous block of code. Can you help me understand how this works?

Reply:
The WaitGroup here is for everything waiting on the session, which in this case is two things: the instance itself calls session.Wait and then this builder.Run code calls session.Wait on the same session. The instance code in this PR waits on the WaitGroup to complete and then signals via the doneChan.

Reply:
So, I misread the code. Sorry for the confusion.

That said, I'm still not sure I'm on board with this. It's very odd to see a user of `Session` reaching into it to update an explicit WaitGroup and then call Wait. Most people would expect the wait to take care of all of that.